libraries/tlv-account-resolution/src/seeds.rs

```diff
@@ -0,0 +1,320 @@
+//! Types for managing seed configurations in TLV Account Resolution
+//!
+//! The largest possible seed configuration is 3 bytes (`InstructionArg`).
+//! This means that an `AccountMetaPda` can store up to 10 seed configurations.
+
+use {crate::error::AccountResolutionError, solana_program::program_error::ProgramError};
+
+/// Enum to describe a required seed for a Program-Derived Address
+#[derive(Clone, Debug, PartialEq)]
+pub enum Seed {
+    /// Uninitialized configuration byte space
+    Uninitialized,
+    /// A literal hard-coded argument
+    Literal {
+        /// The literal value repesented as a vector of bytes.
+        /// For example, if a literal value is a string literal,
+        /// such as "my-seed", this value would be
+        /// `"my-seed".as_bytes().to_vec()`.
+        bytes: Vec<u8>,
+    },
+    /// An instruction-provided argument, to be resolved from the instruction
+    /// data
+    InstructionArg {
+        /// The index where the bytes of an instruction argument begin
+        index: u8,
+        /// The length of the instruction argument (number of bytes)
+        ///
+        /// Note: Max seed length is 32 bytes, so `u8` is appropriate here
+        length: u8,
+    },
+    /// The public key of an account from the entire accounts list.
+    /// Note: This includes an extra accounts required.
+    AccountKey {
+        /// The index of the account in the entire accounts list
+        index: u8,
+    },
+}
+impl Seed {
+    /// Get the size of a seed configuration
+    pub fn tlv_size(&self) -> u8 {
+        match &self {
+            // 1 byte for the discriminator
+            Self::Uninitialized => 0,
+            // 1 byte for the discriminator, 1 byte for the length of the bytes, then the raw bytes
+            Self::Literal { bytes } => 1 + 1 + bytes.len() as u8,
+            // 1 byte for the discriminator, 1 byte for the index, 1 byte for the length
+            Self::InstructionArg { .. } => 1 + 1 + 1,
+            // 1 byte for the discriminator, 1 byte for the index
+            Self::AccountKey { .. } => 1 + 1,
+        }
+    }
+
+    /// Packs a seed configuration into a slice
+    pub fn pack(&self, dst: &mut [u8]) -> Result<(), ProgramError> {
+        if dst.len() != self.tlv_size() as usize {
+            return Err(AccountResolutionError::NotEnoughBytesForSeed.into());
+        }
+        if dst.len() > 32 {
+            return Err(AccountResolutionError::SeedConfigsTooLarge.into());
+        }
+        match &self {
+            Self::Uninitialized => {
+                dst[0] = 0;
+            }
```

Comment:
Yeah this one had me scratching my head a bit. It's like: "should we stop you from packing a zero using the `Uninitialized` variant?" 

Unpacking doesn't care if it encounters a stray 0 between valid seed config entries, so I didn't think this could do any harm (?)

I wish there was a way to make one variant private...

Reply:
unpacking is definitely fine, and even necessary! but we probably want to avoid packing them